### ISSUE TITLE
Unable to totally clear a translated field while Fluent is installed

### ISSUE THREAD

**Issue by phptek** [tractorcow/silverstripe-fluent#204]
No matter what I try, nor what Fluent commit I revert to, I cannot clear the value of a translated field via the CMS.

### Env

FlientLocale Cookie: en_NZ
SilverStripe 3.2.1
Fluent: 3.4.0 (Also tried master)
PHP 5.6 / Ubuntu 14.04
MySQL 5.5.49-0ubuntu0.14.04.1
Extensions applied:
  SiteTree:
    - FluentFilteredExtension
    - MyFluentExtension

1) Edit a translated field in the CMS for a page, Save & Publish = OK.

    mysql> SELECT ID,Fax,Fax_en_NZ FROM Page WHERE ID = 1208 UNION SELECT   ID,Fax,Fax_en_NZ FROM yPage_Live WHERE ID = 1208;

    +------+----------+-----------+
    | ID   | Fax      | Fax_en_NZ |
    +------+----------+-----------+
    | 1208 | 55512345 | 55512345  |
    +------+----------+-----------+

2) Edit a translated field again in the CMS for a page, down to a single character, Save & Publish = OK.

    mysql> SELECT ID,Fax,Fax_en_NZ FROM Page WHERE ID = 1208 UNION SELECT ID,Fax,Fax_en_NZ FROM Page_Live WHERE ID = 1208;

    +------+----------+-----------+
    | ID   | Fax      | Fax_en_NZ |
    +------+----------+-----------+
    | 1208 | 5        | 5         |
    | 1208 | 55512345 | 55512345  |
    +------+----------+-----------+

3) Edit a translated field again in the CMS for a page, but remove ALL characters (empty the field), Save & Publish = FAIL. 

The fields "Fax" and "Fax_en_NZ" on Page_Live both seem to retain the value of the most recent edit (From last entry in Page_versions AFAIKT) but draft table ("Page") shows correct NULL value for the "Fax" field. No matter how often this is repeated, the Page_Live field can never seem to be truncated via the CMS.

    mysql> SELECT ID,Fax,Fax_en_NZ FROM Page WHERE ID = 1208 UNION SELECT ID,Fax,Fax_en_NZ FROM Page_Live WHERE ID = 1208;

    +------+------+-----------+
    | ID   | Fax  | Fax_en_NZ |
    +------+------+-----------+
    | 1208 | NULL | 5         |
    | 1208 | 5    | 5         |
    +------+------+-----------+

### Scenarios

1. [FAIL] Remove / comment all project-specific fluent mods, YML config etc, keeping Fluent 3.4.0
2. [OK] Remove fluent itself
3. [FAIL] Put fluent back, upgrade to dev-master
4. [FAIL] Checkout suspect Fluent commit: 5856da69b14
5. [FAIL] Checkout suspect Fluent commit: e49dccb7066
6. [FAIL] Checkout suspect Fluent commit: 4a69d193df7
7. [OK] Remove fluent
8. [FAIL] Put Fluent back in again, sans custom config of any kind

**Comment by tractorcow** [comment]
Hi @phptek let's debug this. :) Can you please add the additional info:
- Which locales are you using?
- Which is the default?
- If you're adding MyFluentExtension, does THAT extend FluentExtension (or FluentSiteTreeExtension?) If you have fluent applied twice it breaks everything.
- Which locales are you attempting to remove the value in? in the default, or in non default?

**Comment by tractorcow** [comment]
Just to be fair to you, fluent is very "self-correctey" about null values. In some cases it really tries it's best to find a non-null value if one is encountered... so maybe that needs adjusting.

If you have a non-null value in the default, it's not expected behaviour to allow your "translation" to be null. However, you should theoretically be able to set the default to empty, and then translations should allow empty values after the fact.

**Comment by tractorcow** [comment]
+------+------+-----------+
    | ID   | Fax  | Fax_en_NZ |
    +------+------+-----------+
    | 1208 | NULL | 5         |
    | 1208 | 5    | 5         |
    +------+------+-----------+
    
If en_NZ is the default locale, then Fax and Fax_en_NZ should always have the same value. I'm guessing it's not the default?

**Comment by phptek** [comment]
> Which locales are you using?

    Fluent:
      restrict_native_lang: true
      locales:
        - en_NZ
        - ar_EG
        - zh_CN
        - zh_TW
        - nl_NL
        - fr_FR
        - de_DE
        - id_ID
        - it_IT
        - ja_JP
        - ko_KR
        - pl_PL
        - pt_BR
        - ru_RU
        - es_ES
        - tr_TR
        - vi_VN

> Which is the default?

    i18n:
      default_locale: en_NZ

> If you're adding MyFluentExtension, does THAT extend FluentExtension

    class MyFluentExtension extends SiteTreeExtension {

And in config.yml

    SiteTree:
      extensions:
        - FluentFilteredExtension
        - MyFluentExtension

> Which locales are you attempting to remove the value in? in the default, or in non default?

Good question: The default.

**Comment by tractorcow** [comment]
OK, I think that's enough information. :D

So if you are in locale en_NZ and set at field to '', I would expect that to write to both `Field` and `Field_en_NZ`, but it appears to be only written to `Field`.

**Comment by purplespider** [comment]
I've just hit this issue too, but when using the non-default locale.

````
default_locale: en_GB
locales:
    - en_GB
    - fr_FR
    - nl_NL
````
If I try to empty any `fr_FR` field in the CMS, clicking Save doesn't delete the content, it stays as it was.

Rather than refuse to delete, I wouldn't mind if it populated the field with the default local's content so that I had a basis to re-translate, either that or it should at least let me remove the content.

**Comment by phptek** [comment]
@purplespider 

> Rather than refuse to delete, I wouldn't mind if it populated the field with the default local's content so that I had a basis to re-translate

The problem with that approach is that it would never be possible for CMS authors to _totally remove_ a translation for a given locale. The system obviously can have no idea what language the text is "physically" written-in, just that _some text_ exists. In this example there would always be _some text_, ergo the system would always think there was _some translation_.

**Comment by purplespider** [comment]
@phptek 
> The problem with that approach is that it would never be possible for CMS authors to totally remove a translation for a given locale.

Yeh, I only suggested it in case the current implementation which doesn't allow a field to be emptied was done for a reason, but it does sound like it is a bug then.

**Comment by phptek** [comment]
@purplespider @tractorcow I only just spotted #156 which sounds damn near identical to this problem. Could someone verify? If so, I'll close this one in favour of #156

**Comment by purplespider** [comment]
I don't actually think #156 is the exact same issue that I'm experiencing, it seems to report that when trying to empty a field for a translation, the field is populated with a different locale, but for me, it just re-populates the field with the exact same locale content that I was trying to delete.

The issues are probably located, but don't seem to be the same.

**Comment by sanderha** [comment]
I have this problem as well. Was this ever fixed?

**Comment by phptek** [comment]
@sanderha No it hasn't. @tractorcow (The Maintainer) would definitely have closed the ticket had it been resolved. I embarked on a workaround which would selectively "neutralize" problematic fields (See the W.I.P "Fix" linked directly above your comment).

There are at least 2 other issues that seem to relate to the same problem - one workaround (hack) I'm currently using, uses a customised `__get()` call in `Page`. See 3rd comment from bottom in #156

**Comment by tractorcow** [comment]
It's a real bug that isn't quite fixed. I hope I can get some time to fix it soon.

The bug is on writing the manipulation.

**Comment by tractorcow** [comment]
@frankmullenger found an issue in Versioned where he was unable to set a null value.

The fault was an isset() check to see if a field was being set, which should have been an array_key_exists (since setting a null would cause it to fail). Will need to check if it's the same issue here. :)

**Comment by dio5** [comment]
Same problem here. And it's pretty much of a showstopper. Has a fix for this been found already?

**Comment by mleutenegger** [comment]
As mentioned by @tractorcow before, the issue seems to actually be an isset() check which should be an array_key_exists(). As it is quite a frequently occurring issue in a productive environment, I (think I have) fixed it with #224
